contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -563,6 +588,188 @@ pltsql_GetNewObjectId(VariableCache variableCache)
 	ShmemVariableCache->oidCount = 0;
 }
 
+static Oid
+pltsql_GetNewTempObjectId()
+{
+	Oid			result;
+	Oid			tempOidStart;
+	static Oid 	nextTempOid = InvalidOid;
+
+	/* safety check, we should never get this far in a HS standby */
+	if (RecoveryInProgress())
+		elog(ERROR, "cannot assign OIDs during recovery");
+
+	/*
+	 * temp_oid_buffer_size = 0 would indicate that the feature is 
+	 * disabled, so we shouldn't even reach this code.
+	 */
+	Assert(temp_oid_buffer_size > 0);
+
+	if (!OidIsValid(BUFFER_START_TO_OID)) /* InvalidOid means it needs assigning */
+	{
+		/* First check to see if another connection has already picked a start, then update. */
+		LWLockAcquire(OidGenLock, LW_EXCLUSIVE);
+		if (OidIsValid(ShmemVariableCache->tempOidStart))
+		{
+			/*
+			 * Persist the newfound value of temp_oid_buffer_start to disk (via pg_settings).
+			 *
+			 * If for whatever reason this fails, we will fallback to manually updating the GUC,
+			 * which won't be crash-resilient, but won't cause loss of functionality.
```

Comment:
As mentioned before, if we don't persist then we could potentially be incrementing ShmemVariableCache->nextOid by TEMP_OID_BUFFER_SIZE each time it crashes. It would not actually consume OIDs of course, but it might induce wraparound earlier if crashes are happening consistently. Either way, that would be a departure from the design document. I think if we want to consider a change now it's too late.